internal/cluster/connection_invocation_handler.go

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerror"
+	"github.com/hazelcast/hazelcast-go-client/internal/invocation"
+	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
+)
+
+type ConnectionInvocationHandlerCreationBundle struct {
+	ConnectionManager *ConnectionManager
+	ClusterService    *ServiceImpl
+	Logger            ilogger.Logger
+}
+
+func (b ConnectionInvocationHandlerCreationBundle) Check() {
+	if b.ConnectionManager == nil {
+		panic("ConnectionManager is nil")
+	}
+	if b.ClusterService == nil {
+		panic("ClusterService is nil")
+	}
+	if b.Logger == nil {
+		panic("Logger is nil")
+	}
+}
+
+type ConnectionInvocationHandler struct {
+	connectionManager *ConnectionManager
+	clusterService    *ServiceImpl
+	smart             int32
+	logger            ilogger.Logger
+	cb                *cb.CircuitBreaker
+}
+
+func NewConnectionInvocationHandler(bundle ConnectionInvocationHandlerCreationBundle) *ConnectionInvocationHandler {
+	bundle.Check()
+	// TODO: make circuit breaker configurable
+	circuitBreaker := cb.NewCircuitBreaker(
+		cb.MaxRetries(3),
+		cb.MaxFailureCount(3),
+		cb.RetryPolicy(func(attempt int) time.Duration {
+			return time.Duration(attempt) * time.Second
+		}))
+	return &ConnectionInvocationHandler{
+		connectionManager: bundle.ConnectionManager,
+		clusterService:    bundle.ClusterService,
+		logger:            bundle.Logger,
+		cb:                circuitBreaker,
+	}
+}
+
+func (h *ConnectionInvocationHandler) Invoke(inv invocation.Invocation) error {
+	_, err := h.cb.Try(func(ctx context.Context) (interface{}, error) {
+		if h.clusterService.SmartRoutingEnabled() {
```

Comment:
It looks like just to be able to check this we are adding a dispatcher to partition service and setting a boolean and removing the dispatcher back. 
Do we really need all these ceremony ?
If the operation is partition based invokeSmart will fail after seeing that partition owner is nil. And this is not something expensive. 

Also note that for address based operations this `SmartRoutingEnabled` gives to wrong result. 
Address based operations can go to an address directly even-though partition table is not arrived yet.

Reply:
Thanks, simplified smart routing and fixed invoke at: https://github.com/yuce/hazelcast-go-client/commit/7fae2fb2c319ead787a4968cb6047e6f53e87d20 and https://github.com/yuce/hazelcast-go-client/commit/002631f06968e580970d0adb7966ea7d45d59495